ServerCore/ServerMessages/PresenceStore.cs

```diff
@@ -88,10 +88,52 @@ public async Task MergePresenceState(PresenceMessage[] allPresence)
                 await OnPresenceChange(message);
             }
         }
+    }
+
+    /// <summary>
+    /// Model for a user's presence on a puzzle page
+    /// </summary>
+    public class PresenceModel
+    {
+        public int UserId { get; set; }
+        public string Name { get; set; }
+        public PresenceType PresenceType { get; set; }
+    }
 
-        private class TeamStore
+    /// <summary>
+    /// Has the pages present for all puzzles on a team
+    /// </summary>
+    public class TeamStore
+    {
+        public ConcurrentDictionary<int, TeamPuzzleStore> TeamPuzzleStores { get; } = new ConcurrentDictionary<int, TeamPuzzleStore>();
+
+        /// <summary>
+        /// Event for a person joining or leaving any team puzzle. Aggregates all team puzzles.
+        /// </summary>
+        public event Func<int, IList<PresenceModel>, Task> OnTeamPresenceChange;
+
+        internal TeamPuzzleStore GetOrCreateTeamPuzzleStore(int puzzleId)
         {
-            public ConcurrentDictionary<int, TeamPuzzleStore> TeamPuzzleStores { get; } = new ConcurrentDictionary<int, TeamPuzzleStore>();
+            TeamPuzzleStore teamPuzzleStore = new TeamPuzzleStore(puzzleId);
+            if (TeamPuzzleStores.TryAdd(puzzleId, teamPuzzleStore))
+            {
+                teamPuzzleStore.OnTeamPuzzlePresenceChange += InvokeTeamPresenceChange;
+            }
+            else
+            {
+                teamPuzzleStore = TeamPuzzleStores[puzzleId];
+            }
+
+            return teamPuzzleStore;
+        }
+
+        private async Task InvokeTeamPresenceChange(int puzzleId, IList<PresenceModel> presentUsers)
+        {
+            var onTeamPuzzlePresenceChange = OnTeamPresenceChange;
+            if (onTeamPuzzlePresenceChange != null)
+            {
+                await onTeamPuzzlePresenceChange?.Invoke(puzzleId, presentUsers);
```

Comment:
unnecessary ? since 133 already nullchecks and nothing can change the value

Reply:
Fixed